googledataprocauthenticator/test_google_auth.py

```diff
@@ -0,0 +1,411 @@
+
+def test_initialize_credentials_with_auth_dropdown_default_credentials_to_default_credentials():
+    """If Google Authenticator is initialized with default-credentials, if the account dropdown is not
+    changed to a different account, credentials should not be reinitialized when the endpoint is added."""
+    with patch('subprocess.check_output', return_value=auth_list), \
+    patch('google.auth.default', return_value=(not_refreshed_credentials(), 'project')) as d, \
+    patch('google.auth._cloud_sdk.get_auth_access_token', return_value='token'), \
+    patch('google.oauth2._client.refresh_grant', return_value=('token', 'refresh', \
+    expiry, grant_response)), \
+    patch('sparkmagic.auth.google.list_credentialed_accounts', return_value=mock_credentialed_accounts):
+        google_auth = GoogleAuth()
+        assert_equals(google_auth.active_credentials, 'default-credentials')
+        google_auth.initialize_credentials_with_auth_account_selection(google_auth.active_credentials)
+        assert_equals(google_auth.active_credentials, 'default-credentials')
+        d.assert_has_calls([call(scopes=google_auth.scopes), call(scopes=google_auth.scopes)])
+
+def test_initialize_credentials_with_auth_dropdown_user_credentials_to_user_credentials():
+    """If Google Authenticator is initialized with user credentials, if the account dropdown is not
+    changed to a different account, credentials should not be reinitialized when the endpoint is added."""
+    with patch('subprocess.check_output', return_value=AUTH_DESCRIBE_USER), \
+    patch('google.auth.default', side_effect=DefaultCredentialsError), \
+    patch('google.auth._cloud_sdk.get_auth_access_token', return_value='token'), \
+    patch('google.oauth2._client.refresh_grant', return_value=('token', 'refresh', \
+    expiry, grant_response)), \
+    patch('sparkmagic.auth.google.list_credentialed_accounts', return_value=mock_credentialed_accounts):
+        google_auth = GoogleAuth()
+        print(google_auth_class.list_credentialed_accounts)
+        assert_equals(google_auth.active_credentials, 'account@google.com')
+        google_auth.initialize_credentials_with_auth_account_selection(google_auth.active_credentials)
+        google.auth.default.assert_called_once_with(scopes=google_auth.scopes)
+    
+
+@raises(RetryError)
+def test_generate_component_gateway_url_raises_retry_error():
+    with patch('google.cloud.dataproc_v1beta2.ClusterControllerClient.get_cluster', \
+        side_effect=RetryError('error message', 'cause')):
+        google_auth_class.get_component_gateway_url('project', 'region', 'cluster', make_credentials())
+
+@raises(GoogleAPICallError)
+def test_generate_component_gateway_url_raises_google_api_error():
+    with patch('google.cloud.dataproc_v1beta2.ClusterControllerClient.get_cluster', \
+        side_effect=GoogleAPICallError('error message')):
+        google_auth_class.get_component_gateway_url('project', 'region', 'cluster', make_credentials())
+
```

Comment:
Nit: two more tests worth adding:
1) Making sure the correct component gateway url is returned in a successful get_cluster request.
2) Making sure one of correct component gateway urls is returned when listing a pool

Reply:
How would I mock a successful get_cluster request. The get_cluster response returns a response which then I do `url = response.config.endpoint_config.http_ports.popitem()[1]` which gives me a random value (string thats a url) from the httpPorts map. https://cloud.google.com/dataproc/docs/reference/rest/v1/ClusterConfig#endpointconfig So I need to mock the response so that I can call `response.config.endpoint_config.http_ports.popitem()[1]` but what I have tried so far doesn't seem to like the nested attributes so I am not sure how to mock the response. Is there a better way I should do this?